FE-1154 - [Domains UX Redesign] List page - Make tables filterable

nicklemmon opened the pull request:
[FE-1154](https://sparkpost.atlassian.net/browse/FE-1154)

### What Changed

- Adds filtering behavior for the Sending, Bounce, and Tracking domains table for the Domains feature re-design
- Adds new, re-usable hook `useTable` hook that helps with the creation of composable table and table-like interfaces:
  + Adds supporting tests
  + Adds Storybook documentation
- Adds new `Table.SortButton` component for use with `useTable`
- Adds new index file to `src/hooks` so that hooks can be imported like so:
```javascript
import { useTable } from 'src/hooks'
```
- Adds new, re-usable component `Divider`
- Fixes flaky reporter builder Cypress test that was failing the build - updates a component in reporter builder to use correct underline HTML semantics from an a11y perspective

### How To Test

- Go to `/domains/list/sending`, `/domains/list/bounce`, and `/domains/list/tracking` and verify the tables filter as expected
- Do you like the API for `useTable`? if you don't, let's talk about it!

### To Do

- [ ] Incorporate feedback
- [ ] Cut ticket to Introduce debounce in to table filter
- [ ] Cut ticket to add pagination support to `useTable`
- [ ] Cut ticket to refactor fugly `filterByCollectionValues` array helper
- [ ] Replace "Failed Verification" tag with "Unverified"
- [x] Fix tracking domains table implementation now that `useTable` is available
- [x] Write automated tests around `useTable`
- [x] Date not stable locally vs. in CI
- [x] Fix double border in domain status popover
- [x] Content between buttons needs more space
- [x] Remove "Select All" checkbox
- [x] When no subaccount is on a row, render "Primary account"
- [x] When no date is present render "Creation date: unknown"
- [x] Reduce font size for subaccount and creation date using design token provided by UX: **fontSize.200, lineHeight.200**
- [x] Add unit tests for updated selectors
- [x] Some funky state problems when rapidly switching between tabs - disable while `pending`?
- [x] Handle table sorting via `useTable`
- [x] Handle table filtering via `useTable`
- [x] Add button content population for checkbox filters
- [x] Add unit tests for `Table.SortButton` and `Table.SortIcon`
- [x] Write docs for `useTable` with examples
- [x] Refactor `SendingDomainsTable` to reduce its length - probably have a `SendingDomainsTab` component as well
- [x] Fix sorting such that initially selected filters apply
- [x] The layout on smaller screen sizes is a little rough
- [x] Fix rendering "subaccount default" bounce domains
- [x] Fix simultaneous filtering and sorting
- [x] Incorporate `<fieldset>` and `<legend>` for the domain status checkboxes
- [x] Cypress tests for:
  + [x] Sending domains filtering/sorting
  + [x] Bounce domains filtering/sorting
  + [x] Tracking domains filtering/sorting

nicklemmon commented on src/helpers/array.js:
```diff
@@ -1,6 +1,54 @@
-  const conjunction = (array.length >= 3) ? `, ${conjuctionTerm} ` : ` ${conjuctionTerm} `;
-  return array.slice(0, -2).join(', ') + //combines all elements except the last 2 into comma separated items.
+  const conjunction = array.length >= 3 ? `, ${conjuctionTerm} ` : ` ${conjuctionTerm} `;
+  return (
+    array.slice(0, -2).join(', ') + //combines all elements except the last 2 into comma separated items.
     (array.slice(0, -2).length ? ', ' : '') + //appends comma before appending last 2 items unless there are <=2 item in the whole array
-    (array.slice(-2).join(conjunction)); //joins the last 2 items using oxford comma for >=3 items and append to end of sentence.
+    array.slice(-2).join(conjunction)
+  ); //joins the last 2 items using oxford comma for >=3 items and append to end of sentence.
+};
```
Just Prettier changes here.

nicklemmon commented on src/helpers/array.js:
```diff
@@ -1,6 +1,54 @@
+ * @param {*} filters - Set of filters applied to the passed in array.
+ * @example // TODO: Flesh this out
+ */
+export const filterByCollectionValues = (arr, { filters }) => {
```
This is the new stuff.

nicklemmon replied:
...and it's fugly. I would like to refactor this to make it more functional, i.e., removing the use of `.forEach`

nicklemmon commented on src/components/labels/LabelSpacer.js:
```diff
@@ -1,16 +1,23 @@
+function LabelSpacer({ className }) {
   return (
-    <LabelWrapper aria-hidden="true">
+    <LabelWrapper aria-hidden="true" className={className}>
```
Allows style composition via `styled-components`

bkemper commented on src/components/divider/Divider.js:
```diff
@@ -0,0 +1,6 @@
+import { Box } from 'src/components/matchbox';
+
+export default function Divider() {
+  return <Box as="hr" margin="0" />;
```
We already have `<LineBreak />` is there a way we could merge the two or just use it?

nicklemmon replied:
`<LineBreak/>` does something slightly different - and it's only used in one spot that should get phased out in a UX redesign!

bkemper commented on src/components/divider/Divider.stories.mdx:
```diff
@@ -0,0 +1,25 @@
+
+# Divider
+
+The `<Divier/>` renders a styled `<hr/>` element without extraneous margin.
```
👏

bkemper commented on src/components/labels/test/LabelSpacer.test.js:
```diff
@@ -6,14 +6,21 @@ import LabelSpacer from '../LabelSpacer';
+  });
+
+  it('renders with the passed in `className`', () => {
+    useHibana.mockImplementationOnce(() => [{ isHibanaEnabled: true }]);
```
I need to post something in our channel today, but I added this flag to `<TestApp />`.

```
const { container } = render(
  <TestApp isHibanaEnabled={true}>
    <LabelSpacer className="hello-there" />
  </TestApp>
);
```

nicklemmon replied:
of course! That's so much nicer. Let me use that.

bkemper commented on src/helpers/tests/array.test.js:
```diff
@@ -25,3 +25,91 @@ describe('.toSentence', () => {
+      ],
+    });
+
+    expect(arr).toStrictEqual([william]);
```
We should use this more.

nicklemmon replied:
yeah, took me awhile to find it. Super useful.

bkemper commented on src/selectors/trackingDomains.js:
```diff
@@ -16,53 +16,78 @@ export const convertStatus = ({ verified, compliance_status }) => {
+      subaccountName: subaccount_name,
+      subaccountId: subaccount_id,
+      blocked: resolvedStatus === 'blocked',
+      failedVerification: resolvedStatus === 'unverified',
```
Is "unverified" the "failed" status?  I believe "unverified" could mean that they haven't yet tried to verify.

nicklemmon replied:
hmm, yeah it looks like we can't really tell the difference between "Failed Verification" and "Unverified". I'll bring it up with UX.

bkemper commented on src/selectors/sendingDomains.js:
```diff
@@ -1,61 +1,90 @@
+      readyForBounce: readyFor?.bounce,
+      defaultBounceDomain: is_default_bounce_domain,
+      readyForDKIM: readyFor?.dkim,
+      validSPF: status?.spf_status === 'valid',
```
To add to this, `resolveStatus` and `resolveReadyFor` both receive `status` as an argument and immediate deconstruct it.  So, if `status` was undefined, then both of those function calls above would fail before we get to this.

I recommend removing the optional chaining.

bkemper commented on src/pages/reportBuilder/components/AddFilterLink.js:
```diff
@@ -32,6 +32,7 @@ export const AddFilterLink = ({
   const handleClick = e => {
     if (!isNewTabKeyPressed) {
       addFilters([newFilter]);
+      // eslint-disable-next-line no-unused-expressions
```
Does event need to be optional?

nicklemmon replied:
🤷 I wouldn't think so? I just added the ESLint disable comment - not sure why my machine picked up on this error but not Kevin's?

bkemper reviewed (commented):
I left a couple of notes and I cut a couple tickets.

- I cut FE-1184 for the missing "Filters state should be reflected by query params/be deep-linkable" acceptance criteria.
- I cut FE-1183 to address "Explore resolving text truncation issue with subaccounts on this page while here" note

bkemper commented on src/components/matchbox/Table.js:
```diff
@@ -62,14 +98,55 @@ function TotalsRow(props) {
+SortButton.propTypes = {
+  onClick: PropTypes.func.isRequired,
+  children: PropTypes.node.isRequired,
+  sortDirection: PropTypes.oneOf(['asc', 'desc', undefined]),
```
Is `undefined` necessary?

nicklemmon replied:
Well, it can be assorted by ascending, descending, or neither - and when I didn't have this defined, `PropTypes` was throwing an error. Is there a better way to address that? Might be.

bkemper commented on src/helpers/array.js:
```diff
@@ -1,6 +1,54 @@
+  if (hasEmptyFilters) return arr;
+
+  let relevantFilters = [];
+  let filterConfig = {};
```
nit: these should be `const` since you are never reassigning the variable, just mutating it.

bkemper commented on src/helpers/array.js:
```diff
@@ -1,6 +1,54 @@
+
+    arr.forEach(item => {
+      if (item.hasOwnProperty(filter.name)) {
+        return relevantFilters.push(filter);
```
nit: This should be returned.

bkemper commented on src/hooks/useTable/useTable.js:
```diff
@@ -0,0 +1,83 @@
+  sortDirection: undefined,
+};
+
+function useTable(data = []) {
```
👏

bkemper commented on src/pages/domains/ListPage.js:
```diff
@@ -58,85 +44,35 @@ export default function DomainsPage() {
-                    prefix={<Search />}
-                  />
-                </Column>
+          <div>
```
`<div />` or `<Box />`?

nicklemmon replied:
I tend to just use `<div>` when that's all I need - and `<Box/>` when I need access to the `styled-system` API. You thinking that "our" `div` should just be a `<Box/>`? Basically avoid most raw HTML inside features?

bkemper commented on src/pages/domains/components/TrackingDomainsTable.js:
```diff
@@ -1,117 +1,84 @@
+  if (blocked) return <Tag color="red">Blocked</Tag>;
 
-      {status === 'unverified' && <Tag color="yellow">Failed Verification</Tag>}
+  if (failedVerification) return <Tag color="yellow">Failed Verification</Tag>;
```
I mentioned this in the selector, but is this accurate?  "unverified" could just mean that they haven't clicked the verify button yet.

nicklemmon replied:
Let me follow up with UX - I think you're right - but we don't have a mock for "unverified" vs. "failedVerification".

nicklemmon commented:
@bkemper Ready for another look when you have a chance - addressed your feedback in a few spots. There are a couple of things I need to follow up with UX on. Have a meeting with them this afternoon to do so.

bkemper commented:
@nicklemmon After you resolve the merge conflicts, we are good to merge.
